helix-term/src/commands.rs

```diff
@@ -3594,6 +3594,38 @@ fn normal_mode(cx: &mut Context) {
 
     doc.mode = Mode::Normal;
 
+    let doc_changes = doc.changes().changes();
+    let text = doc.text().slice(..);
+    let pos = doc.selection(view.id).primary().cursor(text);
+    let mut can_restore_indent = false;
+
+    // Will remove whitespaces when user opens a new line(may additional insert only tabs or spaces)
+    // and goto normal mode.
+    //
+    // NOTE: when open a new line, changes will only contains 3 elements:
+    // retain(pos) -> insert(something) -> retain(remaining_length), the last retain is useless, we
+    // only concerned about the first two changes.
+    use helix_core::Operation;
+    if let [Operation::Retain(move_pos), Operation::Insert(ref inserted_str), Operation::Retain(_)] =
+        doc_changes
+    {
+        if move_pos + inserted_str.len32() as usize == pos
+            && inserted_str.starts_with('\n')
+            && inserted_str.chars().all(|ch| ch.is_whitespace())
```

Comment:
Needs to use https://github.com/helix-editor/helix/blob/92c2d5d3bffb52a165fad75b6db2120de37b5ae4/helix-core/src/chars.rs#L35-L62

Reply:
Done